Update devices to have minimum 2K RAM and heap, also added test

c1728p9 opened the pull request:
Update the linker scripts (memory map) of targets to have atleast  2K free ram and 2K free heap. 
Add a test to ensure the same, this test should be the first test that fails due to running out of ram or heap.

c1728p9 committed:
Add a minimum requirements HAL test

Add a test to ensure that devices have at least 2K free ram
and 2K free heap. This test should be the first test that fails
due to running out of ram or heap.

theotherjimmy reviewed (approved):
Approved by @bulislaw

deepikabhavnani commented:
@ARMmbed/mbed-os-maintainers  - Can we trigger build on this PR, wanted to check if we can fix the current offending devices.

cmonr commented:
Only running build for now.

/morph build

deepikabhavnani commented:
@c1728p9 @bulislaw  @maciejbocianski - Please review

c1728p9 commented:
Looks good to me

jeromecoutant commented:
Hi

Here is test result with STM32F0 targets:

| target                | platform_name | test suite                          | result | elapsed_time (sec) | copy_method |
|-----------------------|---------------|-------------------------------------|--------|--------------------|-------------|
| NUCLEO_F070RB-ARM     | NUCLEO_F070RB | tests-mbed_hal-minimum_requirements | OK     | 16.89              | default     |
| NUCLEO_F070RB-GCC_ARM | NUCLEO_F070RB | tests-mbed_hal-minimum_requirements | FAIL   | 18.17              | default     |
| NUCLEO_F070RB-IAR     | NUCLEO_F070RB | tests-mbed_hal-minimum_requirements | FAIL   | 16.71              | default     |
| NUCLEO_F072RB-ARM     | NUCLEO_F072RB | tests-mbed_hal-minimum_requirements | OK     | 16.94              | default     |
| NUCLEO_F072RB-GCC_ARM | NUCLEO_F072RB | tests-mbed_hal-minimum_requirements | FAIL   | 18.14              | default     |
| NUCLEO_F072RB-IAR     | NUCLEO_F072RB | tests-mbed_hal-minimum_requirements | FAIL   | 16.78              | default     |
| NUCLEO_F091RC-ARM     | NUCLEO_F091RC | tests-mbed_hal-minimum_requirements | OK     | 16.96              | default     |
| NUCLEO_F091RC-GCC_ARM | NUCLEO_F091RC | tests-mbed_hal-minimum_requirements | OK     | 18.33              | default     |
| NUCLEO_F091RC-IAR     | NUCLEO_F091RC | tests-mbed_hal-minimum_requirements | OK     | 16.89              | default     |

deepikabhavnani commented:
Is that result with just test file or entire PR, I had updated IAR files for Nucleo_F072RB and Nucleo_F070RB. Can update GCC as well.

jeromecoutant commented:
> Is that result with just test file or entire PR

git checkout origin/master
Then applied full PR patch

````
Applied patch TESTS/mbed_hal/minimum_requirements/main.cpp cleanly.
Applied patch targets/TARGET_Freescale/TARGET_KLXX/TARGET_KL25Z/device/TOOLCHAIN_IAR/MKL25Z4.icf cleanly.
Applied patch targets/TARGET_NUVOTON/TARGET_NANO100/device/TOOLCHAIN_IAR/NANO130.icf cleanly.
Applied patch targets/TARGET_STM/TARGET_STM32F0/TARGET_NUCLEO_F070RB/device/TOOLCHAIN_IAR/stm32f070xb.icf cleanly.
Applied patch targets/TARGET_STM/TARGET_STM32F0/TARGET_NUCLEO_F072RB/device/TOOLCHAIN_IAR/stm32f072xb.icf cleanly.
Applied patch targets/TARGET_TOSHIBA/TARGET_TMPM066/device/TOOLCHAIN_IAR/tmpm066fwug.icf cleanly.
Applied patch targets/TARGET_TOSHIBA/TARGET_TMPM3H6/device/TOOLCHAIN_IAR/tmpm3h6fwfg.icf cleanly.
warning: 1 line applied after fixing whitespace errors.
````

Commit:
Set ISR stack and heap sizes for low end devices

Set the ISR stack to be 1KB. https://github.com/ARMmbed/mbed-os/pull/7238
Set the heap size to 3KB(2KB + overhead + spare) so that atleast 2KB free ram is
available for testing.
With dynamic heap size, explicit size is not required. IAR 7.8 supports
static heap, hence the change is needed in IAR linker files.

deepikabhavnani commented:
Corrected the heap sizes, should be more than 2K (overhead, allocation in boot process .. etc_) if we need guaranteed 2K buffer in test.

Also updated the main thread size to 3K for constrained devices. We do not need huge stack in main thread if we cannot have even 1K in heap/data sections.

jeromecoutant commented:
New result with last commit:

| target                | platform_name | test suite                          | result | elapsed_time (sec) | copy_method |
|-----------------------|---------------|-------------------------------------|--------|--------------------|-------------|
| NUCLEO_F070RB-ARM     | NUCLEO_F070RB | tests-mbed_hal-minimum_requirements | OK     | 16.69              | default     |
| NUCLEO_F070RB-GCC_ARM | NUCLEO_F070RB | tests-mbed_hal-minimum_requirements | FAIL   | 17.92              | default     |
| NUCLEO_F070RB-IAR     | NUCLEO_F070RB | tests-mbed_hal-minimum_requirements | OK     | 16.65              | default     |
| NUCLEO_F072RB-ARM     | NUCLEO_F072RB | tests-mbed_hal-minimum_requirements | OK     | 16.64              | default     |
| NUCLEO_F072RB-GCC_ARM | NUCLEO_F072RB | tests-mbed_hal-minimum_requirements | FAIL   | 17.88              | default     |
| NUCLEO_F072RB-IAR     | NUCLEO_F072RB | tests-mbed_hal-minimum_requirements | OK     | 17.55              | default     |
| NUCLEO_F091RC-ARM     | NUCLEO_F091RC | tests-mbed_hal-minimum_requirements | OK     | 16.68              | default     |
| NUCLEO_F091RC-GCC_ARM | NUCLEO_F091RC | tests-mbed_hal-minimum_requirements | OK     | 18.1               | default     |
| NUCLEO_F091RC-IAR     | NUCLEO_F091RC | tests-mbed_hal-minimum_requirements | OK     | 16.66              | default     |

jeromecoutant commented on targets/targets.json:
```diff
@@ -890,7 +891,7 @@
         },
         "detect_code": ["0755"],
-        "macros_add": ["CMSIS_VECTAB_VIRTUAL", "CMSIS_VECTAB_VIRTUAL_HEADER_FILE=\"cmsis_nvic.h\""],
+        "macros_add": ["CMSIS_VECTAB_VIRTUAL", "CMSIS_VECTAB_VIRTUAL_HEADER_FILE=\"cmsis_nvic.h\"", "MBED_CONF_APP_MAIN_STACK_SIZE=3072"],
```
Strange way to define MBED_CONF_APP_xxx macro ?
Usually, it is more defined as config overrides ?
or this should be defined in mbed_rtx.h file ?

jeromecoutant replied:
See #8106

deepikabhavnani replied:
CONF options is already available and used for application overrides. In this case we need target specific value, `mbed_rtx.h` looks like good candidate. Can check that

deepikabhavnani replied:
@jeromecoutant  - Please share the failure log for GCC_ARM, it was successful with local build for me :-(

deepikabhavnani commented:
@jeromecoutant - Need some help here to understand the GCC_ARM linker files for ST devices. malloc for small chunk of memory in bytes failed for Nucleo_F072RB, below is my observation

As per map file - Heap size is 0x0
```
heap           0x20002ee0        0x0
                0x20002ee0                __end__ = .
                0x20002ee0                end = __end__
 *(.heap*)
                0x20002ee0                __HeapLimit = .

.stack_dummy    0x20002ee0        0x0
 *(.stack*)
                0x20004000                __StackTop = (ORIGIN (RAM) + LENGTH (RAM))
                0x20004000                _estack = __StackTop
                0x20004000                __StackLimit = (__StackTop - SIZEOF (.stack_dummy))
                0x20004000                PROVIDE (__stack, __StackTop)
```

Confirmed same from code `__end__` and `__HeapLimit` were same.

Was comparing linker files and most of the ST devices have heap section as (COPY) and no memory allocated. https://github.com/ARMmbed/mbed-os/blob/88eb8aab2330c9ab2b8948079f123ba4b777224a/targets/TARGET_STM/TARGET_STM32F0/TARGET_NUCLEO_F072RB/device/TOOLCHAIN_GCC_ARM/STM32F072XB.ld#L133

jeromecoutant commented:
Hi

I propose to remove OS5 capability for NUCLEO_F070RB and NUCLEO_F072RB.

deepikabhavnani commented:
GCC failure is because newlib  `malloc` allocates memory in 4K chunks and we do not have 4K aligned memory since ISR stack is at the end of RAM, we actually loose 3K memory as a result of this. 
https://github.com/ARMmbed/mbed-os/issues/5386

Commit:
Set main thread stack size as 3K for constrained targets

Commit:
Reposition heap at the end of RAM to be 4K aligned

HEAP memory should be 4K aligned for GCC newlib, with ISR stack at the end of
RAM memory we loose 3K of RAM memory. This fix is for device with <16K RAM to
use RAM entirely.

deepikabhavnani commented:
@jeromecoutant  - Please review 4bb5836be7b4300fa097cf22fe91c154f66f614d. I have moved HEAP to the end of RAM memory. I believe we can have exceptions for special targets < 16K RAM as we loose 3K RAM with existing linker scripts.

jeromecoutant commented:
Hi
With the patch the binary files becomes very big (more than 40 MB..)

deepikabhavnani commented:
> With the patch the binary files becomes very big (more than 40 MB..)

My bad was working with IDE and didn't realize the binary size. Fixed now

jeromecoutant commented on targets/TARGET_STM/TARGET_STM32F0/TARGET_NUCLEO_F072RB/device/TOOLCHAIN_GCC_ARM/NUCLEO_F072RB_retarget.c:
```diff
@@ -0,0 +1,55 @@
+/* mbed Microcontroller Library
+ * Copyright (c) 2018, STMicroelectronics
```
Could you rename the file name to be more generic ? and then copy it to NUCLEO_F070 also ?

deepikabhavnani replied:
Sure, I can add the fix for  "NUCLEO_F070" as well, wanted to confirm first it is fine or not.

deepikabhavnani replied:
Done

jeromecoutant reviewed (commented):
Single test is OK now.
Let me start a full non regression during week end!

deepikabhavnani commented:
Re-build try to see if any other targets failing the build for this test.

/morph build

deepikabhavnani commented:
One more try to see if build is successful now

/morph build

deepikabhavnani commented:
@ARMmbed/mbed-os-maintainers - Fixing build issues for targets (linker file and some more changes), using build CI since nothing in queue. Feel free to kill the test if its launched

NirSonnenschein commented:
/morph export-build

maciejbocianski reviewed (approved):
Tested on nucleo_f070rb on all compilers and all profiles and looks good

jeromecoutant commented:
NUCLEO_F070RB test:

| NUCLEO_F070RB-GCC_ARM | NUCLEO_F070RB | tests-mbedmicro-rtos-mbed-heap_and_stack                                     | FAIL        | 18.81              | default     |

Test was OK with 5.10.0

deepikabhavnani commented:
@jeromecoutant - Please share the test log. I dont have NUCLEO_F070RB :-( , but have NUCLEO_F072RB.

jeromecoutant commented:
Same result with NUCLEO_F072RB, test is failing here :

https://github.com/ARMmbed/mbed-os/blob/c8734cf62a0d9a99b0906ae7500276b3f8c70ca6/TESTS/mbedmicro-rtos-mbed/heap_and_stack/main.cpp#L209

cmonr commented:
@jeromecoutant Would toy mind providing steps to reproduce?

For myself, running the following all succeeded without issues:
```
git clone https://github.com/ARMmbed/mbed-os
cd mbed-os
git fetch origin refs/pull/5285/head:pr5285
git checkout pr5285
mbed test --compile -t GCC_ARM -m NUCLEO_F072RB
```

jeromecoutant commented:
@cmonr , maybe I was not explicit enough.
With this fix, **every** tests are OK,
including the new test from this PR,
but not tests-mbedmicro-rtos-mbed-heap_and_stack

So I would appreciate to check this regression before merging PR

Thx

maciejbocianski commented:
Did some additional tests with NUCLEO_F070RB and `tests-mbedmicro-rtos-mbed-heap_and_stack` with the below code. With pure master and master + the changes from this PR

I can see following problems:
- change didn't swap heap and isr_stack as stated in 045d46571de039e2834915041d704faccfc94d12
- change cause that real isr stack is not at `mbed_stack_isr_start` (thus `test_isr_stack_in_range` is failing)
- swapping heap and isr_stack violates Mbed memory model (https://os.mbed.com/docs/v5.10/reference/memory.html)

```c++
diff --git a/TESTS/mbedmicro-rtos-mbed/heap_and_stack/main.cpp b/TESTS/mbedmicro-rtos-mbed/heap_and_stack/main.cpp
index d4334ad6f..b72daec4a 100644
--- a/TESTS/mbedmicro-rtos-mbed/heap_and_stack/main.cpp
+++ b/TESTS/mbedmicro-rtos-mbed/heap_and_stack/main.cpp
@@ -54,8 +54,70 @@ struct linked_list {
     uint8_t data[MALLOC_TEST_SIZE];
 };

+#define MEM_LOG

+#ifdef MEM_LOG
+volatile uint32_t isr_variable1_addr = 0;
+volatile uint32_t isr_variable2_addr = 0;
+volatile uint32_t isr_variable3_addr = 0;

+void fun()
+{
+    uint8_t isr_variable1 = 1;
+    uint8_t isr_variable2 = 2;
+    uint8_t isr_variable3 = 3;
+    isr_variable1_addr = (uint32_t)&isr_variable1;
+    isr_variable2_addr = (uint32_t)&isr_variable2;
+    isr_variable3_addr = (uint32_t)&isr_variable3;
+
+}
+
+void test_log()
+{
+    mbed_rtos_storage_thread_t *thread = (mbed_rtos_storage_thread_t *) osThreadGetId();
+
+    uint32_t psp = __get_PSP();
+    uint8_t *stack_mem = (uint8_t *) thread->stack_mem;
+    uint32_t stack_size = thread->stack_size;
+    uint8_t main_stack_variable1 = 3;
+    uint8_t main_stack_variable2 = 2;
+    uint8_t main_stack_variable3 = 3;
+
+    utest_printf("mbed_main_stack_size: %X\r\n", stack_size);
+    utest_printf("%X: mbed_main_stack_start\r\n", (uint32_t)stack_mem);
+    utest_printf("%X: psp\r\n", psp);
+    utest_printf("%X: main_stack_variable1_addr\r\n", (uint32_t)&main_stack_variable1);
+    utest_printf("%X: main_stack_variable2_addr\r\n", (uint32_t)&main_stack_variable2);
+    utest_printf("%X: main_stack_variable3_addr\r\n", (uint32_t)&main_stack_variable3);
+    utest_printf("%X: mbed_main_stack_end\r\n", (uint32_t)stack_mem + stack_size);
+
+    utest_printf("mbed_heap_size: %X\r\n", mbed_heap_size);
+    utest_printf("%X: mbed_heap_start\r\n", (uint32_t)mbed_heap_start);
+    uint8_t *heap_variable1_addr = (uint8_t*)malloc(1);
+    utest_printf("%X: heap_variable1_addr\r\n", (uint32_t)heap_variable1_addr);
+    uint8_t *heap_variable2_addr = (uint8_t*)malloc(1);
+    utest_printf("%X: heap_variable2_addr\r\n", (uint32_t)heap_variable2_addr);
+    uint8_t *heap_variable3_addr = (uint8_t*)malloc(1);
+    utest_printf("%X: heap_variable3_addr\r\n", (uint32_t)heap_variable3_addr);
+    utest_printf("%X: mbed_heap_end\r\n", (uint32_t)mbed_heap_start + mbed_heap_size);
+    free(heap_variable1_addr);
+    free(heap_variable2_addr);
+    free(heap_variable3_addr);
+
+    utest_printf("mbed_stack_isr_size: %X\r\n", mbed_stack_isr_size);
+    utest_printf("%X: mbed_stack_isr_start\r\n", mbed_stack_isr_start);
+
+    Timeout t;
+    t.attach_us(fun, 1);
+    while (isr_variable3_addr == 0);
+    utest_printf("%X: isr_variable1_addr\r\n", isr_variable1_addr);
+    utest_printf("%X: isr_variable2_addr\r\n", isr_variable2_addr);
+    utest_printf("%X: isr_variable3_addr\r\n", isr_variable3_addr);
+    uint32_t msp = __get_MSP();
+    utest_printf("%X: msp\r\n", msp);
+    utest_printf("%X: mbed_stack_isr_end\r\n", (uint32_t)mbed_stack_isr_start + mbed_stack_isr_size);
+
+
+}
+#endif
 /* TODO: add memory layout test.
  *
  * The test was skipped for now since not all devices seems to comply with Mbed OS memory.
@@ -234,6 +296,9 @@ void test_heap_allocation_free(void)

 // Test cases
 Case cases[] = {
+#ifdef MEM_LOG
+    Case("test_log", test_log),
+#endif
     Case("Test heap in range", test_heap_in_range),
     Case("Test main stack in range", test_main_stack_in_range),
     Case("Test isr stack in range", test_isr_stack_in_range),

```

Results:

```
no fix (pure master)

mbed_main_stack_size: 1000
20001B90: mbed_main_stack_start
20002A80: psp
20002A8D: main_stack_variable1_addr
20002A8E: main_stack_variable2_addr
20002A8F: main_stack_variable3_addr
20002B90: mbed_main_stack_end


mbed_heap_size: DF0
20002E10: mbed_heap_start
20002E20: heap_variable1_addr
20002E30: heap_variable2_addr
20002E40: heap_variable3_addr
20003C00: mbed_heap_end


mbed_stack_isr_size: 400
20003C00: mbed_stack_isr_start
20003F75: isr_variable1_addr
20003F76: isr_variable2_addr
20003F77: isr_variable3_addr
20003FC0: msp
20004000: mbed_stack_isr_end
```

```
with fix

mbed_main_stack_size: C00
20001F90: mbed_main_stack_start
20002A80: psp
20002A8D: main_stack_variable1_addr
20002A8E: main_stack_variable2_addr
20002A8F: main_stack_variable3_addr
20002B90: mbed_main_stack_end


mbed_heap_size: DF0
20002E10: mbed_heap_start
20002E20: heap_variable1_addr
20002E30: heap_variable2_addr
20002E40: heap_variable3_addr
20003C00: mbed_heap_end


mbed_stack_isr_size: 400
20003C00: mbed_stack_isr_start
20000FF5: isr_variable1_addr
20000FF6: isr_variable2_addr
20000FF7: isr_variable3_addr
20001040: msp
20004000: mbed_stack_isr_end
```

jeromecoutant commented:
Just a comment about tested fix: it is maybe now far from the original PR... "Add a minimum requirements HAL test"
Maybe this PR has to be split into STM32 fix and new test addition ?

deepikabhavnani commented:
Fixed `tests-mbedmicro-rtos-mbed-heap_and_stack`, defines `ISR_STACK_START` and `ISR_STACK_SIZE` were missed out, setting them to actual stack values instead defaults.

@jeromecoutant - Fix is not just for STM targets, but for all devices with low RAM/ROM memories hence will like to keep it in same PR as it is related and will help in tracking specific targets fixed for this requirement. Also commits are different, and I will update the header as well.

jeromecoutant commented:
Hi
Is it possible to rebase now that #8013 has been merged ?
Thx

maciejbocianski commented:
After last fix `mbed-heap_and_stack` test is OK

```
+-----------------------+---------------+------------------------------------------+-------------------------------+--------+--------+--------+--------------------+
| target                | platform_name | test suite                               | test case                     | passed | failed | result | elapsed_time (sec) |
+-----------------------+---------------+------------------------------------------+-------------------------------+--------+--------+--------+--------------------+
| NUCLEO_F070RB-GCC_ARM | NUCLEO_F070RB | tests-mbedmicro-rtos-mbed-heap_and_stack | Test heap allocation and free | 1      | 0      | OK     | 0.11               |
| NUCLEO_F070RB-GCC_ARM | NUCLEO_F070RB | tests-mbedmicro-rtos-mbed-heap_and_stack | Test heap in range            | 1      | 0      | OK     | 0.04               |
| NUCLEO_F070RB-GCC_ARM | NUCLEO_F070RB | tests-mbedmicro-rtos-mbed-heap_and_stack | Test isr stack in range       | 1      | 0      | OK     | 0.06               |
| NUCLEO_F070RB-GCC_ARM | NUCLEO_F070RB | tests-mbedmicro-rtos-mbed-heap_and_stack | Test main stack in range      | 1      | 0      | OK     | 0.05               |
| NUCLEO_F070RB-GCC_ARM | NUCLEO_F070RB | tests-mbedmicro-rtos-mbed-heap_and_stack | test_log                      | 1      | 0      | OK     | 0.64               |
+-----------------------+---------------+------------------------------------------+-------------------------------+--------+--------+--------+--------------------+

```

cmonr commented:
/morph build
